src/schema/type_and_directive_specification.rs

```diff
@@ -0,0 +1,871 @@
+                    TypeKind::from(existing_type)
+                )));
+            };
+
+            let existing_value_set: IndexSet<Name> = existing_type
+                .values
+                .iter()
+                .map(|val| val.0.clone())
+                .collect();
+            let actual_value_set: IndexSet<Name> =
+                self.values.iter().map(|val| val.name.clone()).collect();
+            if existing_value_set != actual_value_set {
+                let enum_type_name = &self.name;
+                let expected_value_names: Vec<String> = existing_value_set
+                    .iter()
+                    .map(|name| name.to_string())
+                    .collect();
+                let actual_value_names: Vec<String> = actual_value_set
+                    .iter()
+                    .map(|name| name.to_string())
+                    .collect();
+                return Err(SingleFederationError::TypeDefinitionInvalid {
+                    message: format!("Invalid definition of type {enum_type_name}: expected values [{}] but found [{}].",
+                    expected_value_names.join(", "), actual_value_names.join(", "))
+                }.into());
+            }
+            return Ok(());
+        }
+
+        let type_pos = EnumTypeDefinitionPosition {
+            type_name: self.name.clone(),
+        };
+        type_pos.pre_insert(schema)?;
+        type_pos.insert(
+            schema,
+            Node::new(EnumType {
+                description: None,
+                name: type_pos.type_name.clone(),
+                directives: Default::default(),
+                values: self
+                    .values
+                    .iter()
+                    .map(|val| {
+                        (
+                            val.name.clone(),
+                            Component::new(EnumValueDefinition {
+                                description: val.description.as_ref().map(|s| s.into()),
+                                value: val.name.clone(),
+                                directives: Default::default(),
+                            }),
+                        )
+                    })
+                    .collect(),
+            }),
+        )
+    }
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// DirectiveSpecification
+
+#[derive(Clone)]
+pub(crate) struct DirectiveArgumentSpecification {
+    pub base_spec: ArgumentSpecification,
+    pub composition_strategy: Option<ArgumentCompositionStrategy>,
+}
+
+type ArgumentMergerFn = dyn Fn(&str, &[Value]) -> Value;
+
+pub(crate) struct ArgumentMerger {
+    pub merge: Box<ArgumentMergerFn>,
+    pub to_string: Box<dyn Fn() -> String>,
+}
+
+type ArgumentMergerFactory =
+    dyn Fn(&FederationSchema) -> Result<ArgumentMerger, SingleFederationError>;
+
+pub(crate) struct DirectiveCompositionSpecification {
+    pub supergraph_specification: fn(federation_version: Version) -> Box<dyn SpecDefinition>,
```

Comment:
The closures here might be a bit too difficult to use once we start working on actual composition, but we can fix that some time in the future \o/